### ISSUE TITLE
jax.numpy.linspace() does not faithfully recover endpoints

### ISSUE THREAD

**Issue by jakevdp** [google/jax#3014]
Example:
```python
>>> import numpy as np
>>> import jax.numpy as jnp
>>> a, b = -2.3328583, 2.8459014
>>> np.linspace(a, b, 10)[-1] == b
True
>>> jnp.linspace(a, b, 10)[-1] == b
DeviceArray(False, dtype=bool)
```

**Comment by jakevdp** [comment]
Looks to be due to float32 roundoff error:
```
>>> b - a, float32(b) - float32(a)
(5.1787598, 5.1787596)
```

**Comment by jakevdp** [comment]
...and numpy side-steps this by doing all computation in 64-bit, even if input and output dtypes are 32-bit: https://github.com/numpy/numpy/blob/fc1f196584ff6dd530982febba2322679317c632/numpy/core/function_base.py#L120-L121

**Comment by jakevdp** [comment]
I think we could address this by factoring out operations to prevent loss of precision; e.g.
```python
def linspace_bad(a, b, N):
  a, b = np.float32(a), np.float32(b)
  return a + (b - a) * np.arange(N, dtype=np.float32)

def linspace_good(a, b, N):
  a, b = np.float32(a), np.float32(b)
  step = np.arange(N, dtype=np.float32) / (N - 1)
  return a * (1 - step) + b * step

print(linspace_bad(-2.3328583, 2.8459015, 2))
print(linspace_good(-2.3328583, 2.8459015, 2))
```
```
[-2.3328583  2.8459013]
[-2.3328583  2.8459015]
```

**Comment by mattjj** [comment]
Wow, amazing find!
